### PR TITLE
[SPARK-35881][SQL][FOLLOWUP] Add a boolean flag in AdaptiveSparkPlanExec to ask for columnar output

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/33140 to propose a simpler idea for integrating columnar execution into AQE.

Instead of making the `ColumnarToRowExec` and `RowToColumnarExec` dynamic to handle `AdaptiveSparkPlanExec`, it's simpler to let the consumer decide if it needs columnar output or not, and pass a boolean flag to `AdaptiveSparkPlanExec`.

For Spark vendors, they can set the flag differently in their custom columnar parquet writing command when the input plan is `AdaptiveSparkPlanExec`.

One argument is if we need to look at the final plan of AQE and consume the data differently (either row or columnar format). I can't think of a use case and I think we can always statically know if the AQE plan should output row or columnar data.

### Why are the changes needed?

code simplification.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

manual test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:503]: This is the new change. Previously we always assume that the final plan should output row data and add `ColumnarToRowExec` if the plan is columnar. Now the caller side can specify if the final plan should output columnar format or not.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:526]: seems we don't need the post-stage creation extension rules at all if we have an API `ShuffleExchangeLike.requiresColumnarInput` (or reuse `SparkPlan.supportsColumnar`?). We can let the rule `ApplyColumnarRulesAndInsertTransitions` to adjust the plan and always output columnar format if the shuffle/broadcast needs columnar input.

**Comment by cloud-fan** [comment]
cc @andygrove  @tgravescs @dongjoon-hyun

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala
Comment (maryannxue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:503]: Add javadoc?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:503]: Looks like a good idea.

BTW, since `ApplyColumnarRulesAndInsertTransitions` is used a few time, shall we remove the default value?
```
sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:case class ApplyColumnarRulesAndInsertTransitions(
sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:      ApplyColumnarRulesAndInsertTransitions(sparkSession.sessionState.columnarRules),
sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:    ApplyColumnarRulesAndInsertTransitions(context.session.sessionState.columnarRules),
sql/core/src/test/scala/org/apache/spark/sql/execution/ColumnarRulesSuite.scala:    val rules = ApplyColumnarRulesAndInsertTransitions(
sql/core/src/test/scala/org/apache/spark/sql/execution/ColumnarRulesSuite.scala:    val rules = ApplyColumnarRulesAndInsertTransitions(
```

**Comment by tgravescs** [comment]
We will take a look, I'll likely get to it tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (andygrove) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: I like the idea of adding this parameter but I'm not understanding how this would get set to `true`, especially from a Spark plugin.  It looks like new instances of `AdaptiveSparkPlanExec` are only created by the `InsertAdaptiveSparkPlan` rule and that does not explicitly specify a value for `outputColumnar` in this PR. Would a plugin need to replace this rule with a custom version perhaps?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (maryannxue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:346]: Just override `supportsColumnar` in constructor?

**Comment by HyukjinKwon** [comment]
Yeah, It think this approach is fine.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: For example:
```
case class WriteToParquetColumnarExec(input: SparkPlan, ...) extends UnaryExecNode {
  def doExecute... = {
    val finalPlan = input match {
      case a: AdaptiveSparkPlanExec => a.copy(outputColumnar = true)
      case _ => input
    }
    finalPlan.executeColumnar().mapPartitions ...
  }
}
```

Another idea is to add custom physical rules. Currently, we don't have an extension point to allow people to add custom physical rules (that deal with the physical plan produced by the planner). One workaround is to leverage `ColumnarRule.postColumnarTransitions`:
```
class MyColumnarRule... {
  def postColumnarTransitions = new Rule[SparkPlan] {
    def apply(input: SparkPlan) = input match {
      case w @ WriteToParquetColumnarExec(a: AdaptiveSparkPlanExec, ...) =>
        w.copy(input = a.copy(outputColumnar = true))
      case other => other
    }
  }
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: link to the previous discussion: https://github.com/apache/spark/pull/33624#discussion_r682088051

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: Thanks. I really like this approach and this will work for us.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:525]: I am looking at this suggestion next to see how this would impact our usage. If we are to make changes like this I would prefer it if we could do these in a separate PR though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:351]: can we move this method back?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: can we keep the docs of https://github.com/apache/spark/pull/33624/files#diff-ec42cd27662f3f528832c298a60fffa1d341feb04aa1d8c80044b70cbe0ebbfcL199-L203?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:525]: If that works without post-stage creation extension, I would prefer to revert it since Spark 3.2 is not out yet.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:502]: maybe `outputsColumnar` to match with `supportsColumnar` or `isOutputColumnar`?

**Review by HyukjinKwon — APPROVED** [review]
Looks good to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:351]: it's not just moving around, it follows the other methods to also call `withFinalPlanUpdate`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: that doc doesn't apply any more right now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:525]: I can try it with a new PR after this one is merged, and see if we can revert the post-stage creation extension.

Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: I think that we do need to retain the post-stage creation rules. Spark will insert a RowToColumnar in the adaptive plan for the final query stage but we need a way for our plugin to replace that with a GPU-specific version and we can use the extensible post-stage creation rule to do that.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: Can it be done with `ColumnarRule.postColumnarTransitions`?

**Comment by tgravescs** [comment]
overall seems like this should work, but we are testing a few things out so please hold off on merging until we can finish.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (andygrove) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: Yes, it can.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala
Comment (andygrove) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:530]: This comment is incorrect. We are not going back to rows here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:69]: Can we add new docs?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:530]: oh, I'll remove this comment

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by tgravescs** [comment]
+1, this works for us. I'm good with this change.  Note though we need the postStage rules at this point but are investigating other possibilities there.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by cloud-fan** [comment]
The GA fails on R linter and is not related to this PR, I'm merging it to master/3.2, thanks for the review!

**Pull request closed without merging**
